http/src/main/java/io/micronaut/http/CaseInsensitiveMutableHttpHeaders.java

```diff
@@ -0,0 +1,251 @@
+package io.micronaut.http;
+
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.core.convert.ArgumentConversionContext;
+import io.micronaut.core.convert.ConversionService;
+import io.micronaut.core.util.CollectionUtils;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.TreeMap;
+
+/**
+ * A {@link MutableHttpHeaders} implementation that is case-insensitive.
+ *
+ * @author Tim Yates
+ * @since 4.0.0
+ */
+@Internal
+public class CaseInsensitiveMutableHttpHeaders implements MutableHttpHeaders {
+
+    private final TreeMap<String, List<String>> backing;
+    private final ConversionService<?> conversionService;
+
+    /**
+     * Create an empty CaseInsensitiveMutableHttpHeaders.
+     *
+     * @param conversionService The conversion service
+     */
+    public CaseInsensitiveMutableHttpHeaders(ConversionService<?> conversionService) {
+        this(new HashMap<>(), conversionService);
+    }
+
+    /**
+     * Create a CaseInsensitiveMutableHttpHeaders populated by the entries in the provided {@literal Map<String,String>}.
+     *
+     * @param defaults The defaults
+     * @param conversionService The conversion service
+     */
+    public CaseInsensitiveMutableHttpHeaders(Map<String, List<String>> defaults, ConversionService<?> conversionService) {
+        this.conversionService = conversionService;
+        this.backing = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
+        defaults.forEach((key, value) -> value.forEach(v -> this.add(key, v)));
+    }
+
+    @Override
+    public List<String> getAll(CharSequence name) {
+        List<String> values = backing.get(name.toString());
+        if (CollectionUtils.isEmpty(values)) {
+            return Collections.emptyList();
+        }
+        return values;
+    }
+
+    @Nullable
+    @Override
+    public String get(CharSequence name) {
+        List<String> strings = backing.get(name.toString());
+        if (CollectionUtils.isEmpty(strings)) {
+            return null;
+        }
+        return strings.get(0);
+    }
+
+    @Override
+    public Set<String> names() {
+        return backing.keySet();
+    }
+
+    @Override
+    public Collection<List<String>> values() {
+        return backing.values();
+    }
+
+    @Override
+    public <T> Optional<T> get(CharSequence name, ArgumentConversionContext<T> conversionContext) {
+        String value = get(name);
+        if (value != null) {
+            return conversionService.convert(value, conversionContext);
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public MutableHttpHeaders add(CharSequence header, CharSequence value) {
+        validate(header, value);
+        backing.computeIfAbsent(header.toString(), s -> new ArrayList<>()).add(value.toString());
```

Comment:
Might want to size this list with some sensible default. Repeated headers of the same name are rare and normally there are not many of them

Reply:
I chose 2 as a default 🤔 Maybe 1 is better?